UI/TechTreeWnd.cpp

```diff
@@ -80,22 +81,40 @@ namespace {
         return true;
     }
 
-    struct ToggleCategoryFunctor {
-        ToggleCategoryFunctor(TechTreeWnd* tree_wnd, const std::string& category) : m_tree_wnd(tree_wnd), m_category(category) {}
-        void operator()() { m_tree_wnd->ToggleCategory(m_category); }
+
+    struct SetCategoryViewFunctor {
+        SetCategoryViewFunctor(TechTreeWnd* tree_wnd, const std::string& category) :
+            m_tree_wnd(tree_wnd),
+            m_category(category)
+        {}
+        void operator()(bool b) {
+            if (m_tree_wnd != 0)
+                b ? m_tree_wnd->ShowCategory(m_category) : m_tree_wnd->HideCategory(m_category);
+        }
         TechTreeWnd* const m_tree_wnd;
         const std::string m_category;
     };
 
-    struct ToggleAllCategoriesFunctor {
-        ToggleAllCategoriesFunctor(TechTreeWnd* tree_wnd) : m_tree_wnd(tree_wnd) {}
-        void operator()() { m_tree_wnd->ToggleAllCategories(); }
+    struct SetAllCategoryViewsFunctor {
+        SetAllCategoryViewsFunctor(TechTreeWnd* tree_wnd) :
+            m_tree_wnd(tree_wnd)
+        {}
+        void operator()(bool b) {
+            if (m_tree_wnd != 0)
+                b? m_tree_wnd->ShowAllCategories() : m_tree_wnd->HideAllCategories();
```

Comment:
pushed, as expected these only proceed if the window doesn't exist (?).
The buttons will never trigger the functionality.  Am I missing some other point here?